compiler/pipes/convert-sprintf-calls.cpp

```diff
@@ -0,0 +1,199 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2021 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#include "compiler/pipes/convert-sprintf-calls.h"
+
+struct FormatCallInfo {
+  FormatCallInfo() : args(VertexRange(Vertex::iterator{}, Vertex::iterator{})) {}
+  // stores a variable that contains all the call arguments,
+  // if all arguments are constant
+  VertexAdaptor<op_var> var;
+  // stores an array that contains all the call arguments,
+  // if at least one argument is a function call or a variable
+  VertexAdaptor<op_array> array;
+  // flag for the presence of a value in format_args_var
+  bool is_var{false};
+  // stores a set of vertices of the call arguments from var or array
+  VertexRange args;
+};
+
+struct FormatPart {
+  enum class FormatSpecType {
+    Decimal,
+    String
+  };
+
+  std::string value;
+  FormatSpecType specifier;
+
+  bool is_specifier() const {
+    return value.empty();
+  }
+};
+
+std::vector<FormatPart> try_parse_format_string(const std::string &format) {
+  std::vector<FormatPart> parts;
+  std::string last_value;
+  bool find_percent = false;
+
+  for (const auto &symbol : format) {
+    if (find_percent) {
+      if (symbol == 'd') {
+        parts.push_back(FormatPart{"", FormatPart::FormatSpecType::Decimal});
+      } else if (symbol == 's') {
+        parts.push_back(FormatPart{"", FormatPart::FormatSpecType::String});
+      } else if (symbol == '%') {
+        last_value += "%";
+      } else {
+        return {};
+      }
+      find_percent = false;
+    } else if (symbol == '%') {
+      find_percent = true;
+      if (!last_value.empty()) {
+        parts.push_back(FormatPart{last_value});
+        last_value = "";
+      }
+    } else {
+      last_value += symbol;
+    }
+  }
+
+  if (!last_value.empty()) {
+    parts.push_back(FormatPart{last_value});
+  }
+
+  return parts;
+}
+
+VertexPtr ConvertSprintfCallsPass::on_exit_vertex(VertexPtr root) {
+  if (auto func_call = root.try_as<op_func_call>()) {
+    const auto func = func_call->func_id;
+    if (vk::any_of_equal(func->name, "sprintf", "vsprintf")) {
```

Comment:
What about vprintf, printf, vfprintf, fprintf? The solution this these functions looks the same.

Reply:
It seems to me that this does not make much sense, since the speed for these functions is no critical, and their support will slightly complicate the code and increase the already large diff between versions.

Reply:
but..
replace `f$printf($format, ...)` to `print(string_build)` function call
replace `f$fprintf($stream, $format, ...)` to `f$write($stream, string_build)` function call

But ok, it's up to you..

Reply:
> f$printf($format, ...) can be replaced with print(string_build) function call
ok, this is not truth, `print` always returns 1.. 
anyway, I don't see any problem to add one more simple function, which prints and returns size..

Reply:
> > f$printf($format, ...) can be replaced with print(string_build) function call
> 
> ok, this is not truth, `print` always returns 1..
> anyway, I don't see any problem to add one more simple function, which prints and returns size..

Yep, this is one of the problems with replacement, because `printf` returns the number of characters printed.

You mean something like this:

```cpp
int f$print_with_size(string str) {
   f$print(str);
   return str.size();
}
```

Reply:
> > > f$printf($format, ...) can be replaced with print(string_build) function call
> > 
> > 
> > ok, this is not truth, `print` always returns 1..
> > anyway, I don't see any problem to add one more simple function, which prints and returns size..
> 
> Yep, this is one of the problems with replacement, because `printf` returns the number of characters printed.
> 
> Do you mean something like that in runtime?
> 
> ```c++
> int f$print_with_size(string str) {
>    f$print(str);
>    return str.size();
> }
> ```

Although this can be solved more easily, we can only replace those for which `rl_type` is equal to left.

Reply:
discussed offline: let's leave as is and may be fix it in the future